### PR TITLE
Support specifying output channels in io.image.read_image

### LINKED ISSUE #2948: Support specifying output channels (RGB vs grayscale) for io.image.read_image
## 🚀 Feature
Similar to `channels` argument for `tf.io.decode_image` or `PIL.Image.open(..).convert('format')` allow specifying the output channel 'format' and have a sensible conversion performed for you when decoding an image if set.

## Motivation

Prevent the need to manually check and do color conversion from RGB -> grayscale or grayscale -> RGB when datasets often have mixes of RGB and grayscale images. The output color format of image loading in data pipeline needs to be consistent for many typical use cases.

The lower level libraries (libjpeg, libpng) etc have facilities for specifying the output color space and will do the conversion for you if you set it up (ie `out_color_space` for libjpeg). This is what tensorflow decode_image does and is the most efficient.

## Pitch

Add a `channels=` arg to read_image that matches Tensorflow semantics.

* channels=0 - leave as original, a grayscale image decodes to 1 channel out, rgb to 3
* channels=1 - grayscale out
* channels=3 - RGB out 
* channels=4 - RGBA out (PNG or formats that support alpha, not valid for jpeg)

### PR DESCRIPTION
**Pull request by datumbox**
Fixes #2948

I add a `channels` parameter that allows the users to specify the number of output channels while reading an image. The default value is 0 which leaves the image as-is and ensures the change is BC. The following public API methods  of the `torchvision.io.image` package were updated:

- `decode_png(input, channels=0)`
- `decode_jpeg(input, channels=0)`
- `decode_image(input, channels=0)`
- `read_image(path, channels=0)`

There is a small update on the originally proposed pitch because I added support for grayscale transparency and handling for palette images. Here are the supported values:

- channels=0 - leave as original (grayscale, palette, grayscale with alpha, rgb, rgb with alpha, CMYK etc)
- channels=1 - Grayscale if the image is not palette else leave as Palette
- channels=2 - Grayscale with Alpha (PNG only, not valid for JPEG)
- channels=3 - RGB
- channels=4 - RGB with Alpha (PNG only, not valid for JPEG)

The PR adds 3 JPEG assets with total size 7kb. These are used to test the supported conversions. It also removes a 900kb asset file which is no longer needed. The assets were produced using the following snippet:
```python
from PIL import Image

# manually downloaded from https://pytorch.org/assets/images/pytorch-logo.png
original = 'pytorch-logo.png'
with Image.open(original) as img:
    img = img.convert("RGBA")
    img = img.resize((100, 100)) 
    img.convert("L").save("gray_pytorch.jpg")
    img.convert("RGB").save("rgb_pytorch.jpg")
    img.convert("CMYK").save("cmyk_pytorch.jpg")
```

### CODE REVIEW

File: test/test_cpp_models.py
Comment (datumbox) [inline test/test_cpp_models.py:29]: I had to move the files related to jpeg decoding/writing to their own folder. Some of the tests on `test_image.py` start writing in random locations if they spot a jpeg and this interferes with the changes of this PR. Unfortunately the `grace_hopper_517x606.jpg` is used by a few other tests so here I have to update the locations. See below, my remarks on the problematic tests.

File: test/test_image.py
Comment (datumbox) [inline test/test_image.py:58]: Our previous code on master, allows the loading of CMYK images. Unfortunately libJpeg does not support converting such images to grayscale etc. Thus here I add a test to check that our changes are BC. Unfortunately since many CMYK conversions are not supported, I skip them here.

File: test/test_image.py
Comment (datumbox) [inline test/test_image.py:65]: It seems that this is necessary when comparing CMYK images loaded with PIL vs libjpeg.

File: test/test_image.py
Comment (datumbox) [inline test/test_image.py:77]: Pillow seems to use different weights than libjpeg to convert RGB to grayscale. The differences are slightly visible. This is why here I permit for some variation across images when we convert to grayscale.

File: test/test_image.py
Comment (datumbox) [inline test/test_image.py:175]: My current implementation does not convert palette images to grayscale. This is not a bug/limitation but a feature™. This is because palette images, which can be useful for image segmentation, already have only 1 channel. So when a user specifies `channels=1` there is no need do an additional conversion to grayscale.

File: test/test_image.py
Comment (datumbox) [inline test/test_image.py:185]: Adding a small tolerance on the pixel values. Small variations can be caused by implementation differences between PIL and libpng.

File: test/test_image.py
Comment (datumbox) [inline test/test_image.py:105]: This test is problematic. Once it detects a jpg it assumes it can write to specific locations. This assumption is problematic so I had to move its assets to a separate folder. It's worth providing a more permanent fix for this unit-test on the future.

File: test/test_image.py
Comment (datumbox) [inline test/test_image.py:148]: Another problematic test. See above.

File: torchvision/csrc/cpu/image/readjpeg_cpu.cpp
Comment (datumbox) [inline torchvision/csrc/cpu/image/readjpeg_cpu.cpp:7]: Components or Channels? All JPEG implementations seem to use the word components instead of channels. On the other hand PNG implementations use the word channels.

I ended up using the most accurate word in each implementation but this makes us use the 2 words interchangeably across the codebase, which is definitely not ideal.

File: torchvision/csrc/cpu/image/readjpeg_cpu.cpp
Comment (datumbox) [inline torchvision/csrc/cpu/image/readjpeg_cpu.cpp:115]: libjpeg does not support converting from CMYK to grayscale etc. There is a way to do this but it [involves converting it](https://github.com/tensorflow/tensorflow/blob/86871065265b04e0db8ca360c046421efb2bdeb4/tensorflow/core/lib/jpeg/jpeg_mem.cc#L284-L313) manually to RGB. Perhaps we can do this on the future.

File: torchvision/csrc/cpu/image/readpng_cpu.cpp
Comment (datumbox) [inline torchvision/csrc/cpu/image/readpng_cpu.cpp:80]: There is another type of transparency that we can support. Not sure if it's a must-have for this PR.

File: torchvision/io/image.py
Comment (datumbox) [inline torchvision/io/image.py:140]: Here is the channels/components discussion again.

File: test/test_image.py
Comment (datumbox) [inline test/test_image.py:39]: I am not sure why it was necessary to have the `grace_hopper_517x606.pth` file in the assets. Storing a small jpg image to do reasonable tests is worthwhile sometimes but keeping uncompressed bitmap vectors can quickly increase the size of the library. As far as I can see, we can continue doing the test without storing the pth file. Hence I removed it.

File: test/test_image.py
Comment (datumbox) [inline test/test_image.py:52]: The assets contain a few interesting file combinations and this allows us to confirm that the file conversions work properly.

File: torchvision/csrc/cpu/image/readpng_cpu.cpp
Comment (datumbox) [inline torchvision/csrc/cpu/image/readpng_cpu.cpp:8]: Here we use channels instead of components, see remark on JPEG above.

File: torchvision/io/image.py
Comment (datumbox) [inline torchvision/io/image.py:205]: On user-facing APIs that can be either channels or components, I use channels.

**Review by datumbox — COMMENTED** [review]
I left a few comments to explain parts of the implementation.`

File: test/test_image.py
Comment (datumbox) [inline test/test_image.py:74]: Pillow seems to use different weights than libjpeg to convert RGB to grayscale. The differences are slightly visible but nothing significant. Moreover it seems that on windows, the values don't much 100% either. This is why here I permit for some variation across images.

File: test/test_image.py
Comment (datumbox) [inline test/test_image.py:199]: One can argue that this test is useful because it covers directly the `decode_image()` method. Unfortunately it's flaky and to fix it, I would have to copy-paste the mitigations done on `test_decode_jpeg()`. I decided to remove it because this method is just a wrapper for `decode_jpeg()` and `decode_png()` which are exhaustively tested.

**Comment by datumbox** [comment]
The failing tests on Travis are not related to this PR. I would like to rebase to master once #2985 is merged to ensure all tests still pass.

File: torchvision/csrc/cpu/image/readjpeg_cpu.cpp
Comment (fmassa) [inline torchvision/csrc/cpu/image/readjpeg_cpu.cpp:105]: I see that we are now using `cinfo.num_components` instead of `cinfo.output_components`. Is there any particular difference / reason for this change?

File: torchvision/csrc/cpu/image/readjpeg_cpu.cpp
Comment (fmassa) [reply torchvision/csrc/cpu/image/readjpeg_cpu.cpp:115]: Sounds good to me. Maybe we can add this as a comment in the code?

File: torchvision/csrc/cpu/image/readjpeg_cpu.cpp
Comment (fmassa) [inline torchvision/csrc/cpu/image/readjpeg_cpu.cpp:130]: nit: can we make this a `int64_t`, so that we don't have more warnings during compilation?

File: torchvision/csrc/cpu/image/readpng_cpu.cpp
Comment (fmassa) [reply torchvision/csrc/cpu/image/readpng_cpu.cpp:80]: No need to add this for now I believe

File: torchvision/csrc/cpu/image/readpng_cpu.cpp
Comment (fmassa) [inline torchvision/csrc/cpu/image/readpng_cpu.cpp:144]: don't we need to keep the cast to avoid compiler warnings?

File: torchvision/io/image.py
Comment (fmassa) [reply torchvision/io/image.py:140]: I would really prefer if at least the python API has the same argument names.

File: torchvision/io/image.py
Comment (fmassa) [inline torchvision/io/image.py:221]: Can you mention here that 0 keeps the original number of channels? So that the users at least know what they get by default

File: torchvision/csrc/cpu/image/readpng_cpu.cpp
Comment (fmassa) [inline torchvision/csrc/cpu/image/readpng_cpu.cpp:87]: following my comment just above, I wonder if we shouldn't let `channels=1` to always return a grayscale image.

What do you think?

File: test/test_image.py
Comment (fmassa) [reply test/test_image.py:39]: IIRC this was due to the fact that different PIL versions (or for different OSes) would return slightly different results while decoding the image, so we decided to save an uncompressed image so that we can compare exactly against the ground-truth, without introducing additional uncertainties in test failures in the future if PIL changes something.

File: test/test_image.py
Comment (fmassa) [reply test/test_image.py:74]: > Moreover it seems that on windows, the values don't much 100% either

This is probably the reason why we were using a saved `.pth` for this file. @andfoy might remember a bit more, but I believe that was the reason.

> Pillow seems to use different weights than libjpeg to convert RGB to grayscale

I assume our conversion to RGB in `transforms.functional.convert_rgb_to_grayscale` matches the output that we have in our implementation for grayscale conversion?

File: test/test_image.py
Comment (fmassa) [reply test/test_image.py:182]: Which tests do have a different tolerance? Is it for the grayscale conversion? Maybe we could make the test a bit more strict and only give a tolerance in this case?

File: test/test_image.py
Comment (fmassa) [reply test/test_image.py:65]: Hum, this is interesting / weird. Do you have an understanding of why this is the case? And would you say this is a problem with PIL?

**Review by fmassa — COMMENTED** [review]
Thanks a lot for the PR, it looks great!

I would really prefer if we could unify the name of the arguments across `jpeg` / `png`, this way it is clear to the user that they both represent the same thing.

About `channels=1` keeping the image as a `palette` type, although it makes sense I wonder if the expected behavior would be instead to convert it to a grayscale. Thoughts?

I've made a few other comments on the PR, let me know what you think

File: test/test_image.py
Comment (datumbox) [reply test/test_image.py:65]: I do not have a good understanding of why this happens unfortunately. I just noticed that they are represented differently...

File: test/test_image.py
Comment (datumbox) [reply test/test_image.py:172]: I agree that this complicates the semantics. On the other hand, it does not have a real effect on the situation once it becomes a tensor. 

At any case, I do not mind changing this but it would probably involve a double conversion first to RGB and then to Grayscale because jpeglib does not support a Palette to Grayscale conversion. Thoughts?

File: test/test_image.py
Comment (datumbox) [reply test/test_image.py:182]: If you check original versions of this snippet, you will see that originally I was checking with `allclose` + tolerance only the grayscale for which I knew they have different implementations. Every other conversion was checked for equality. This works find on Linux and macOS but unfortunately not on Windows. This is why I changed this for all. As I'm not on windows I don't know which other conversion fails, but I could experiment with different values and rely on the CI to tell me which works. Thoughts?

File: torchvision/io/image.py
Comment (datumbox) [reply torchvision/io/image.py:140]: I agree, I think I should just replace everything with channels for the shake of simplicity in semantics.

File: test/test_image.py
Comment (datumbox) [reply test/test_image.py:39]: So are we OK to keep the removal and do the check as I'm doing it now or shall I rollback?

**Comment by andfoy** [comment]
The usage of pth files is due to the difference between libjpeg and libjpeg-turbo on Windows and Mac, which right now we are not able to use.

File: test/test_image.py
Comment (datumbox) [reply test/test_image.py:74]: >I assume our conversion to RGB in transforms.functional.convert_rgb_to_grayscale matches the output that we have in our implementation for grayscale conversion?

Precisely we use the same weights:
`L = R * 0.2989 + G * 0.5870 + B * 0.1140`

File: test/test_image.py
Comment (datumbox) [reply test/test_image.py:199]: > Also, can you expand why this is flaky?

For the reasons you covered [here](https://github.com/pytorch/vision/pull/2988#discussion_r522983218). It would require the same mitigation with tolerance I did there. 

Happy to restore it, just note that there will be some duplication of code. An alternative is to replace the use of `decode_jpeg()` with `decode_image()` in the `test_decode_jpeg` unit-test. The code-path will be covered and it will be clear if it breaks what's the issue. Which one do you prefer?

**Comment by datumbox** [comment]
@fmassa Thanks for the review. I marked as resolved anything that I either accept your proposal or is already covered discussed. I kept open anything that requires a second look. I'll send now another commit with the changes. I would appreciate to review the last remaining points.

@andfoy Thanks for providing the background story.

File: test/test_cpp_models.py
Comment (datumbox) [reply test/test_cpp_models.py:29]: Created issue #3002

File: test/test_image.py
Comment (datumbox) [reply test/test_image.py:102]: Should be handled by #3002

File: torchvision/csrc/cpu/image/readjpeg_cpu.cpp
Comment (datumbox) [reply torchvision/csrc/cpu/image/readjpeg_cpu.cpp:7]: done

File: torchvision/csrc/cpu/image/readjpeg_cpu.cpp
Comment (datumbox) [reply torchvision/csrc/cpu/image/readjpeg_cpu.cpp:115]: done

File: torchvision/io/image.py
Comment (datumbox) [reply torchvision/io/image.py:140]: done

File: torchvision/csrc/cpu/image/readjpeg_cpu.cpp
Comment (datumbox) [reply torchvision/csrc/cpu/image/readjpeg_cpu.cpp:105]: Good catch. My understanding is that `num_components` == `out_components` when you use it immediately after a `jpeg_read_header()`. The value of `out_components` changes after you call `jpeg_calc_output_dimensions()`. We also tested these changes with a variety of different assets and it works fine. Nevertheless, I agree that `output_components` requires less assumptions (maybe on the future libjpeg will do a mandatory conversion for some reason), so I'll replace it.

File: torchvision/csrc/cpu/image/readjpeg_cpu.cpp
Comment (datumbox) [reply torchvision/csrc/cpu/image/readjpeg_cpu.cpp]: This is intentional to maintain the same behaviour as before. See the original code: 
- The components and width were integers and so stride was int.
- Since channels/components now comes from Python it's int64_t, so here I define stride as integer explicitly.

Thoughts?

File: torchvision/csrc/cpu/image/readpng_cpu.cpp
Comment (datumbox) [reply torchvision/csrc/cpu/image/readpng_cpu.cpp:144]: No because channels is already an int64_t. No need to copy it.

File: torchvision/io/image.py
Comment (datumbox) [reply torchvision/io/image.py]: Good call.

File: torchvision/csrc/cpu/image/readpng_cpu.cpp
Comment (datumbox) [reply torchvision/csrc/cpu/image/readpng_cpu.cpp:87]: Marking this as resolved this to keep the discussion in one place.

File: torchvision/csrc/cpu/image/readjpeg_cpu.cpp
Comment (datumbox) [reply torchvision/csrc/cpu/image/readjpeg_cpu.cpp:105]: Turns out the proposed change causes issues with damages pictures: https://app.circleci.com/pipelines/github/pytorch/vision/5000/workflows/21a22cd6-0f9f-459b-9c20-b2d1b0ff1c92/jobs/307793

Rolling `output_components ` back to `num_components`.

File: test/test_image.py
Comment (fmassa) [reply test/test_image.py:172]: There is a semantics change in converting it or not to grayscale for the following reason:
some datasets store semantic maps as a palette image, with labels / colors going from 0 to 20 or so. If we were expecting to see the image in `L` format, we would see it with different colors.

Here is one example:
```python
i = Image.open('/datasets01/VOC2007/011719/SegmentationClass/005130.png')
print(i.mode)
# > 'P'

# color values used as labels for semantic segmentation
np.unique(np.array(i))
# > array([  0,   3, 255], dtype=uint8)

# but if we convert them to grayscale for visualization
# values are different
np.unique(np.array(i.convert('L')))
# > array([  0, 113, 220], dtype=uint8)
```

> double conversion first to RGB and then to Grayscale

That's fine with me

Comment (datumbox) [reply test/test_image.py:172]: Sounds good I'll make the change.

File: test/test_image.py
Comment (datumbox) [reply test/test_image.py:199]: Applied my proposed change on latest commit.

File: test/test_image.py
Comment (fmassa) [reply test/test_image.py:65]: I would feel a bit more comfortable if I could understand what is going on here. Is OpenCV giving similar results as PIL for the same case if you pass the appropriate reading flags?

File: test/test_image.py
Comment (fmassa) [reply test/test_image.py:181]: The failures on Windows were probably the same as what we had before, and was probably the reason why we had the `.pth` file

File: test/test_image.py
Comment (fmassa) [reply test/test_image.py:199]: ok. I would still have preferred to keep each function separately testing only one function, as now the `test_decode_*` functions exercise the codepath of two different C++ functions. But the previous tests were not perfect either, so instead of being blocked by this let's move forward

File: test/test_image.py
Comment (fmassa) [reply test/test_image.py:181]: Ideally it would be good to have exact checks here, as we are loading the image, so we should make sure that we are consistently loading and returning the same result across versions.
If by some reason the results are slightly different, it's technically a BC-breaking change and it would be good to catch it warn users about it during the release notes (if we need to break BC)

I don't mind resolving this conversation, but I just wanted to point out that this is probably something we would like to be a bit more strict in the future and move back to exact checks

File: test/test_image.py
Comment (fmassa) [reply test/test_image.py:74]: Same comment as before. I'm resolving this, but any change to torchvision that changes the pixel values returned by it is a BC-breaking change, and we can't detect it now.

File: test/test_image.py
Comment (fmassa) [reply test/test_image.py:39]: Check my other comments.

If we make approximate checks, we can as well introduce silent small BC-breaking changes on the return values of the decoding functions, which won't be caught because the checks are approximate, and we wouldn't be able to warn the users about it

File: test/test_image.py
Comment (datumbox) [reply test/test_image.py:65]: It seems that this investigation is more meaty that I expected, so I raised a separate issue #3012. Note that what you flagged is an existing behaviour and this PR just adds a test that highlights the issue. Hence I propose not to block the PR for this investigation.

Comment (fmassa) [reply test/test_image.py:65]: Sounds good

File: torchvision/csrc/cpu/image/readpng_cpu.cpp
Comment (datumbox) [reply torchvision/csrc/cpu/image/readpng_cpu.cpp:90]: no palette images have colours. See https://www.ffmpeg.org/doxygen/0.5/png_8h.html

**Review by fmassa — APPROVED** [review]
Thanks a lot Vasilis!

**Pull request merged**
